src/backend/oneapi/kernel/gradient.hpp

```diff
@@ -0,0 +1,173 @@
+
+template<typename T>
+using local_accessor = sycl::accessor<T, 1, sycl::access::mode::read_write,
+                                      sycl::access::target::local>;
+template<typename T>
+using read_accessor = sycl::accessor<T, 1, sycl::access::mode::read>;
+template<typename T>
+using write_accessor = sycl::accessor<T, 1, sycl::access::mode::write>;
+
+#define sidx(y, x) scratch_[((y + 1) * (TX_ + 2)) + (x + 1)]
+
+template<typename T>
+class gradientCreateKernel {
+   public:
+    gradientCreateKernel(write_accessor<T> d_grad0, const KParam grad0,
+                         write_accessor<T> d_grad1, const KParam grad1,
+                         read_accessor<T> d_in, const KParam in,
+                         const int blocksPerMatX, const int blocksPerMatY,
+                         const int TX, const int TY, local_accessor<T> scratch)
+        : d_grad0_(d_grad0)
+        , grad0_(grad0)
+        , d_grad1_(d_grad1)
+        , grad1_(grad1)
+        , d_in_(d_in)
+        , in_(in)
+        , blocksPerMatX_(blocksPerMatX)
+        , blocksPerMatY_(blocksPerMatY)
+        , TX_(TX)
+        , TY_(TY)
+        , scratch_(scratch) {}
+    void operator()(sycl::nd_item<2> it) const {
+        auto g = it.get_group();
+
+        const int idz = g.get_group_id(0) / blocksPerMatX_;
+        const int idw = g.get_group_id(1) / blocksPerMatY_;
+
+        const int blockIdx_x = g.get_group_id(0) - idz * blocksPerMatX_;
+        const int blockIdx_y = g.get_group_id(1) - idw * blocksPerMatY_;
+
+        const int xB = blockIdx_x * g.get_local_range(0);
+        const int yB = blockIdx_y * g.get_local_range(1);
+
+        const int tx = it.get_local_id(0);
+        const int ty = it.get_local_id(1);
+
+        const int idx = tx + xB;
+        const int idy = ty + yB;
+
+        const bool cond = (idx >= in_.dims[0] || idy >= in_.dims[1] ||
+                           idz >= in_.dims[2] || idw >= in_.dims[3]);
+
+        int xmax = (TX_ > (in_.dims[0] - xB)) ? (in_.dims[0] - xB) : TX_;
+        int ymax = (TY_ > (in_.dims[1] - yB)) ? (in_.dims[1] - yB) : TY_;
+
+        int iIdx = in_.offset + idw * in_.strides[3] + idz * in_.strides[2] +
+                   idy * in_.strides[1] + idx;
+
+        int g0dx = idw * grad0_.strides[3] + idz * grad0_.strides[2] +
+                   idy * grad0_.strides[1] + idx;
+
+        int g1dx = idw * grad1_.strides[3] + idz * grad1_.strides[2] +
+                   idy * grad1_.strides[1] + idx;
+
+        // local T scratch[(TY_ + 2) * (TX_ + 2)];
+
+        // Multipliers - 0.5 for interior, 1 for edge cases
+        typename std::conditional<std::is_same<T, std::complex<double>>::value,
+                                  double, float>::type
+            xf = 0.5 * (1 + (idx == 0 || idx >= (in_.dims[0] - 1))),
+            yf = 0.5 * (1 + (idy == 0 || idy >= (in_.dims[1] - 1)));
+
+        // Copy data to scratch space
+        T zero = (T)(0);
+        if (cond) {
+            sidx(ty, tx) = zero;
```

Comment:
Not a fan of this macro and indirection. Can this be a function?

Reply:
tricky b/c ty and tx, above, bc of the weird typing... 

```
       using Txy = std::conditional<std::is_same<T, std::complex<double>>::value,
                                  double, float>::type;
         Txy  xf = 0.5 * (1 + (idx == 0 || idx >= (in_.dims[0] - 1)));
          Txy yf = 0.5 * (1 + (idy == 0 || idy >= (in_.dims[1] - 1)));
```

the define is better since it maneuvers out of the situation with far less code and more traceable compiler behavior